src/main/java/org/xstefank/model/PersistentList.java

```diff
@@ -30,32 +32,122 @@ public PersistentList(String dirName, String fileName) {
         this(new File(dirName, fileName));
     }
 
-    public void addUser(String username) {
-        if (list.contains(username)) {
+    @Override
+    public String set(int index, String element) {
+        String oldValue = super.set(index, element);
+        rewriteFileContent(index);
+        return oldValue;
+    }
+
+    @Override
+    public boolean add(String s) {
+        if (contains(s)) {
+            return false;
+        }
+        boolean added = super.add(s);
+        if (added) {
+            try (FileWriter fw = new FileWriter(file, true)) {
+                fw.write(s + LINE_SEPARATOR);
+            } catch (IOException e) {
+                throw new IllegalStateException("Cannot write element to list", e);
+            }
+        }
+        return added;
+    }
+
+    @Override
+    public void add(int index, String element) {
+        if (contains(element)) {
             return;
         }
-        list.add(username);
+        super.add(index, element);
+        rewriteFileContent(index);
+    }
+
+    @Override
+    public String remove(int index) {
+        String element = super.remove(index);
+        rewriteFileContent(index);
+        return element;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        int index = indexOf(o);
+        boolean removed = super.remove(o);
+        if (removed) {
+            rewriteFileContent(index);
+        }
+        return removed;
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
+        try (FileWriter fw = new FileWriter(file, false)) {
+            fw.write("");
+        } catch (IOException e) {
+            throw new IllegalStateException("Cannot clear the list", e);
+        }
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends String> c) {
+        boolean changed = super.addAll(c);
         try (FileWriter fw = new FileWriter(file, true)) {
-            fw.write(username + "\n");
+            for (String element : c) {
+                fw.write(element);
+            }
         } catch (IOException e) {
-            throw new IllegalStateException("Cannot write user to file", e);
+            throw new IllegalStateException("Cannot write elements to list", e);
```

Comment:
shouldn't we guard here against possible inconsistencies? (also maybe in some other methods) -- If collection c contains 3 elements, we successfully write first 2 on line 99 and then IOException is thrown then super has all 3 elements while the file has only 2